src/include/storage/index/bplustree.h

```diff
@@ -0,0 +1,2554 @@
+     * leaf node. For the leaf node where insert occurs, get exclusive access.
+     */
+
+    // Get access to the Tree
+    root_latch_.LockExclusive();
+
+    if (root_ == nullptr) {
+      // If root is nullptr then we make a Leaf Node.
+      KeyNodePointerPair p1, p2;
+      p1.first = element.first;
+      p2.first = element.first;
+      p1.second = nullptr;
+      p2.second = nullptr;
+
+      root_ = ElasticNode<KeyValuePair>::Get(leaf_node_size_upper_threshold_, NodeType::LeafType, 0,
+                                             leaf_node_size_upper_threshold_, p1, p2);
+    }
+
+    // NOTE: At this point, exclusive lock to tree (root_latch_) is held.
+
+    BaseNode *current_node = root_;
+    BaseNode *parent_node = nullptr;
+
+    // Get shared latch on root node
+    current_node->GetNodeSharedLatch();
+
+    // Traversing Down and maintaining a stack of pointers
+    while (current_node->GetType() != NodeType::LeafType) {
+      auto node = reinterpret_cast<ElasticNode<KeyNodePointerPair> *>(current_node);
+
+      // Release parent's shared lock
+      if (parent_node != nullptr) {
+        parent_node->ReleaseNodeSharedLatch();
+      } else {
+        // Release the exclusive lock on the tree
+        root_latch_.UnlockExclusive();
+      }
+
+      // NOTE: FindLocation returns the location of first element that compares greater than
+      auto index_pointer = static_cast<InnerNode *>(node)->FindLocation(element.first, this);
+      // Thus we have to go in the left side of location which will be the
+      // pointer of the previous location.
+      if (index_pointer != node->Begin()) {
+        index_pointer -= 1;
+        parent_node = current_node;
+        current_node = index_pointer->second;
+      } else {
+        parent_node = current_node;
+        current_node = node->GetLowKeyPair().second;
+      }
+
+      // Get current node's shared lock
+      current_node->GetNodeSharedLatch();
+    }
+
+    // Try optimistic insertion into the leaf node, if insertion without splitting is possible.
+
+    // Get current node's exclusive lock and release lock on the parent
+    current_node->ReleaseNodeSharedLatch();
+    current_node->GetNodeExclusiveLatch();
+    if (parent_node != nullptr) {
+      parent_node->ReleaseNodeSharedLatch();
+    } else {
+      root_latch_.UnlockExclusive();
+    }
+
+    // Beyond this we only have exclusive latch on the current_node
+
+    bool finished_insertion = false;
+    // We maintain the element that we have to recursively insert up.
+    // This is the element that has to be inserted into the inner nodes.
+    auto node = reinterpret_cast<ElasticNode<KeyValuePair> *>(current_node);
+
+    auto location_greater_key_leaf = static_cast<LeafNode *>(node)->FindLocation(element.first, this);
+    if (location_greater_key_leaf != node->Begin()) {
+      if (KeyCmpEqual((location_greater_key_leaf - 1)->first, element.first)) {
+        // Key present in tree => insert into value list
```

Comment:
Maybe when talking about the structure of the tree state that we allow duplicate keys